Redesigned the AbstractScreenView class to be Generic

akshaym-webonise opened the pull request:
Reformatted the code.
Redesigned the AbstractScreenView class to be Generic

dheerajwebonise commented on pom.xml:
```diff
@@ -0,0 +1,56 @@
+    </properties>
+
+    <dependencies>
+        <!-- Spring -->
```
Please do not add commented code and if do add TODO statement for it

murtaza-webonise commented on src/main/java/com/webonise/MapSurveyApplication.java:
```diff
@@ -0,0 +1,26 @@
+                ("fxml/MainScreenView.fxml"));
+        primaryStage.setTitle("Map Survey");
+        primaryStage.setMaximized(true);
+        primaryStage.setScene(new Scene(parent, 600, 400));
```
Create Constant for magic numbers

murtaza-webonise commented on src/main/java/com/webonise/MapSurveyApplication.java:
```diff
@@ -0,0 +1,26 @@
+    public void start(Stage primaryStage) throws IOException {
+        Parent parent = FXMLLoader.load(Thread.currentThread().getContextClassLoader().getResource
+                ("fxml/MainScreenView.fxml"));
+        primaryStage.setTitle("Map Survey");
```
Create constant for String

dheerajwebonise commented on src/main/java/com/webonise/MapSurveyApplication.java:
```diff
@@ -0,0 +1,26 @@
+    @Override
+    public void start(Stage primaryStage) throws IOException {
+        Parent parent = FXMLLoader.load(Thread.currentThread().getContextClassLoader().getResource
+                ("fxml/MainScreenView.fxml"));
```
Good! Loading resource from current thread.
Add constants for String

dheerajwebonise commented on pom.xml:
```diff
@@ -0,0 +1,56 @@
+            <artifactId>spring-webmvc</artifactId>
+            <version>${spring.version}</version>
+        </dependency>
+        <dependency>
```
I don't think we need spring web related dependencies here.

dheerajwebonise commented on src/main/java/com/webonise/controllers/MainScreenController.java:
```diff
@@ -0,0 +1,75 @@
+    @FXML
+    private TableColumn lngColumn;
+
+
```
Extra new line

dheerajwebonise commented on src/main/java/com/webonise/controllers/MainScreenController.java:
```diff
@@ -0,0 +1,75 @@
+        coordinates = new Coordinates();
+        webEngine = mapBrowser.getEngine();
+        webEngine.load(getClass().getResource("/html/MapBox.html").toExternalForm());
+        progressBar.progressProperty().bind(webEngine.getLoadWorker().progressProperty());
```
what if loading MapBox.html get failed

dheerajwebonise commented on src/main/java/com/webonise/controllers/MainScreenController.java:
```diff
@@ -0,0 +1,75 @@
+        tableView.setItems(tableRows);
+    }
+
+    public void update(Observable object, Object arg) {
```
Give meaningful names for method parameters

dheerajwebonise commented on src/main/java/com/webonise/controllers/MainScreenController.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.ResourceBundle;
+
+public class MainScreenController extends Region implements Initializable, Observer {
+
```
Controller is tightly coupled with the view

dheerajwebonise replied:
Do we need to implement Oberver here because here we have only one?

dheerajwebonise commented on src/main/java/com/webonise/controllers/MainScreenController.java:
```diff
@@ -0,0 +1,75 @@
+    private WebView mapBrowser;
+    @FXML
+    private ProgressBar progressBar;
+
```
New line. Please be consistent. If you are adding new line for autowiring then here also u should add it.

dheerajwebonise commented on src/main/resources/fxml/MainScreenView.fxml:
```diff
@@ -0,0 +1,33 @@
+                        <TableColumn prefWidth="150" text="Longitude" fx:id="lngColumn"/>
+                    </columns>
+                </TableView>
+                <Button fx:id="clearButton" text="Clear Selection" onAction="#clearTheMap"/>
```
Do not add button actions inside fxml

dheerajwebonise commented on src/main/java/com/webonise/controllers/MainScreenController.java:
```diff
@@ -0,0 +1,75 @@
+    @FXML
+    public void clearTheMap(ActionEvent event) {
+        tableRows.clear();
+        webEngine.executeScript("clearPolygon()");
```
You can write clearPolygon instead of clearPolygon() and please have constants for these Strings

dheerajwebonise commented on src/main/java/com/webonise/controllers/MainScreenController.java:
```diff
@@ -0,0 +1,75 @@
+
+    private static final Logger LOG = LoggerFactory.getLogger(MainScreenController.class);
+
+    private Coordinates coordinates;
```
Coordinate class is annotated with Component then why you dont injected dependency here.

dheerajwebonise commented on src/main/java/com/webonise/models/Coordinates.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.Observable;
+
+@Component
+public class Coordinates extends Observable {
```
If you are going to use this class as POJO then remove annotation Component.
Please override hashCode, equals and toString

dheerajwebonise commented on src/main/resources/js/script.js:
```diff
@@ -0,0 +1,61 @@
+	map = L.mapbox.map('map', 'mapbox.streets').setView([40, -74.50], 3);
+
+    markerCount=1;
+	latlngs = new Array();
```
Declare array with []

dheerajwebonise commented on src/main/resources/logback.xml:
```diff
@@ -0,0 +1,33 @@
+        <appender-ref ref="CONSOLE"/>
+        <appender-ref ref="FILE"/>
+    </root>
+</configuration>
```
Add new line at the end of each and every file

dheerajwebonise commented on src/main/resources/js/script.js:
```diff
@@ -0,0 +1,61 @@
+
+function clearPolygon(){
+    map.removeLayer(polygon);
+    polygon = null;
```
why assigning null here

deepa-webonise commented on src/main/resources/fxml/MainScreenView.fxml:
```diff
@@ -0,0 +1,33 @@
+<HBox xmlns:fx="http://javafx.com/fxml" fx:controller="com.webonise.controllers.MainScreenController" spacing="8">
+    <padding>
+        <Insets left="5.0" right="5.0" top="5.0" bottom="5.0"/>
+    </padding>
```
Remove padding, spacing and height width properties from FXML and use seprate css for that. We dont use inline css into FXML file.

dheerajwebonise commented on src/main/java/com/webonise/controllers/MainController.java:
```diff
@@ -0,0 +1,15 @@
+public class MainController {
+
+    @Autowired
+    MainScreenController mainScreenController;
```
Add access specifiers and make it more restricted

dheerajwebonise commented on src/main/java/com/webonise/controllers/MainScreenController.java:
```diff
@@ -0,0 +1,75 @@
+    private static final Logger LOG = LoggerFactory.getLogger(MainScreenController.class);
+
+    private Coordinates coordinates;
+    private WebEngine webEngine;
```
Can we make member variables thread safe

dheerajwebonise commented on src/main/java/com/webonise/models/Coordinates.java:
```diff
@@ -0,0 +1,50 @@
+@Component
+public class Coordinates extends Observable {
+    private float latitude;
+    private float longitude;
```
Avoid using float.
Floats are not equal safe.
http://stackoverflow.com/questions/1088216/whats-wrong-with-using-to-compare-floats-in-java

dheerajwebonise commented on src/main/resources/html/MapBox.html:
```diff
@@ -0,0 +1,22 @@
+    <script src='https://api.mapbox.com/mapbox.js/v2.2.2/mapbox.js'></script>
+    <script src="https://api.tiles.mapbox.com/mapbox.js/plugins/turf/v2.0.0/turf.min.js"></script>
+    <script src="../js/fireBug.js"></script>
+    <script src='../js/script.js'></script>
```
please have meaningfull name for script.js

dheerajwebonise commented on src/main/resources/html/MapBox.html:
```diff
@@ -0,0 +1,22 @@
+    <script src="../js/fireBug.js"></script>
+    <script src='../js/script.js'></script>
+    <link href='https://api.mapbox.com/mapbox.js/v2.2.2/mapbox.css' rel='stylesheet'/>
+    <style>
```
add style inside css file

dheerajwebonise commented on src/main/resources/js/script.js:
```diff
@@ -0,0 +1,61 @@
+
+ function showPolygon(latlng)
+ {
+ 	if(polygon != undefined)
```
instead you can write if (!polygon)

murtaza-webonise replied:
just write if(polygon) {} because  javascript itself checks for null, undefined and empty you don't need to manually write if(polygon != undefined)

dheerajwebonise commented on src/main/resources/js/script.js:
```diff
@@ -0,0 +1,61 @@
+    console.log(latlng.lat+" "+latlng.lng);
+    coordinates.setLatitude(latlng.lat);
+    coordinates.setLongitude(latlng.lng);
+    coordinates.setDataFromJs(true);
```
why we need this boolean flag?

murtaza-webonise commented on src/main/resources/js/script.js:
```diff
@@ -0,0 +1,61 @@
+ 	L.mapbox.accessToken = 'pk.eyJ1IjoiYWtzbWtyIiwiYSI6ImNpZmY3bnlncDd6eGtzdWtucjFrY3Z3OGkifQ.2QcYNghIAx0ZsZJVewp8QA';
+	map = L.mapbox.map('map', 'mapbox.streets').setView([40, -74.50], 3);
+
+    markerCount=1;
```
try to use prototypical javascript and also format javascript code as well

murtaza-webonise commented on src/main/resources/js/script.js:
```diff
@@ -0,0 +1,61 @@
+var latlngs;
+var polygon;
+var markers;
+var markerCount;
```
If you use prototype or classical inheritance then you don't need to define global variables. Just look into it if you can refactor it

akshaym-webonise commented:
Added the Logger Implementation for javascript.
Fixed the issue with marker deletion
Seperated FXML file and stylesheet for MainScreenView

dheerajwebonise commented:
Use spring for DI

akshaym-webonise commented:
Added the Logger Implementation for javascript.
Fixed the issue with marker deletion
Seperated FXML file and stylesheet for MainScreenView
Implemented Spring Dependency Injection

murtaza-webonise commented on src/main/java/com/webonise/controllers/LatLngController.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    public void updateMarker(LatLng point) {
+        for (int index = 0; index < latLngList.size(); index++) {
```
size should be calculated before loop

murtaza-webonise commented on src/main/java/com/webonise/controllers/LatLngController.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    public void deleteMarker(LatLng point) {
+        for (int index = 0; index < latLngList.size(); index++) {
```
size should be calculated before loop

murtaza-webonise commented on src/main/java/com/webonise/models/LatLng.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
what is o ?? please be particular about naming conventions

dheerajwebonise commented on src/main/java/com/webonise/config/AppConfig.java:
```diff
@@ -0,0 +1,15 @@
+public class AppConfig {
+    @Bean
+    public MainScreenController mainScreenController() {
+        return new MainScreenController();
```
Why we need to create a bean of MainScreenController like this??

murtaza-webonise commented on src/main/java/com/webonise/config/AppConfig.java:
```diff
@@ -0,0 +1,15 @@
+
+@Configuration
+@ComponentScan("com.webonise")
+public class AppConfig {
```
Please use correct name. why you named it as AppConfig ?

murtaza-webonise commented on src/main/java/com/webonise/config/AppConfig.java:
```diff
@@ -0,0 +1,15 @@
+@ComponentScan("com.webonise")
+public class AppConfig {
+    @Bean
+    public MainScreenController mainScreenController() {
```
why you are manually creating bean ? you could have used @Component

dheerajwebonise commented on src/main/java/com/webonise/controllers/LatLngController.java:
```diff
@@ -0,0 +1,48 @@
+@Component
+public class LatLngController {
+
+    private ObservableList<LatLng> latLngList;
```
If you are initializing this list inside constructor then why you didnt make it thread safe?

murtaza-webonise commented on src/main/java/com/webonise/views/AbstractScreenView.java:
```diff
@@ -0,0 +1,30 @@
+            fxmlLoader.setController(this);
+            fxmlLoader.load();
+            LOG.debug("FXML loaded");
+        } catch (IOException e) {
```
again what is e here ?

murtaza-webonise replied:
where is the description of PR ?

dheerajwebonise commented on src/main/java/com/webonise/controllers/LatLngController.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    public ObservableList<LatLng> getLatLngList() {
+        return latLngList;
```
Return unmodifiable list here. Use Collections.unmodifiableList().
Also you may need to provide add/remove methods.

dheerajwebonise commented on src/main/java/com/webonise/controllers/LatLngController.java:
```diff
@@ -0,0 +1,48 @@
+    public void updateMarker(LatLng point) {
+        for (int index = 0; index < latLngList.size(); index++) {
+            LatLng current = latLngList.get(index);
+            if (current.getPointNo() == point.getPointNo()) {
```
float and doubles are not equal safe in java
http://stackoverflow.com/questions/1088216/whats-wrong-with-using-to-compare-floats-in-java

dheerajwebonise commented on src/main/java/com/webonise/controllers/MainScreenController.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Autowired
+    private MainScreenView mainScreenView;
+    private Scene scene;
```
Why we need scene as member variable?

dheerajwebonise commented on src/main/java/com/webonise/controllers/ScriptController.java:
```diff
@@ -0,0 +1,25 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ScriptController {
```
This cant be a Controller

murtaza-webonise replied:
This can't be a controller

dheerajwebonise commented on src/main/java/com/webonise/models/LatLng.java:
```diff
@@ -0,0 +1,76 @@
+public class LatLng {
+    public double lat;
+    public double lng;
+    public int pointNo;
```
pointNo doesnt make sense inside LatLng class

dheerajwebonise commented on src/main/java/com/webonise/models/LatLng.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    @Override
+    public String toString() {
```
Good!

dheerajwebonise commented on src/main/java/com/webonise/views/MainScreenView.java:
```diff
@@ -0,0 +1,84 @@
+public class MainScreenView extends AbstractScreenView {
+
+    private static final Logger LOG = LoggerFactory.getLogger(MainScreenView.class);
+    private static final String SCRIPT_DELETE_MARKERS = "deleteMarkers()";
```
i think we need just the method name and arguments for that method not brackets

dheerajwebonise commented on src/main/java/com/webonise/views/MainScreenView.java:
```diff
@@ -0,0 +1,84 @@
+    private static final String SCRIPT_DELETE_MARKERS = "deleteMarkers()";
+    private static final String SCRIPT_CLEAR_POLYGON = "clearPolygon()";
+    private static final String LAT = "lat";
+    private static final String LNG = "lng";
```
No need to have constants for lat lng

dheerajwebonise commented on src/main/java/com/webonise/views/MainScreenView.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @PostConstruct
+    public void init() {
```
Please refactor this init method

dheerajwebonise replied:
Good

dheerajwebonise commented on src/main/resources/js/mapbox_script.js:
```diff
@@ -0,0 +1,74 @@
+    createPolygon();
+}
+
+var Point = function(latLng, pointNo){
```
Nice

dheerajwebonise commented on pom.xml:
```diff
@@ -0,0 +1,44 @@
+
+    <properties>
+        <version.spring>4.1.0.RELEASE</version.spring>
+        <version.slf4j>1.7.5</version.slf4j>
```
Why we are not using latest spring and slf4j version?

akshaym-webonise commented:
Updated CSS for FXML.
Changed model LatLng to WayPoint
Changed ScriptController to ScriptLogger

murtaza-webonise commented on src/main/java/com/webonise/controllers/WayPointController.java:
```diff
@@ -0,0 +1,50 @@
+@Component
+public class WayPointController {
+
+    private final ObservableList<WayPoint> wayPointList;
```
waypoint is a single word

murtaza-webonise commented on src/main/java/com/webonise/views/MainScreenView.java:
```diff
@@ -0,0 +1,99 @@
+    private void bindButtonActions() {
+        LOG.debug("Binding button actions");
+        clearButton.setOnAction(event -> {
+            webEngine.executeScript(SCRIPT_DELETE_MARKERS);
```
please check webEngine State before calling JS

murtaza-webonise commented:
Closing certain convention needs to be followed and where is description of PR ?

dheerajwebonise commented on src/main/java/com/webonise/views/MainScreenView.java:
```diff
@@ -0,0 +1,118 @@
+        LOG.debug("Initializing webEngine");
+        webEngine = mapBrowser.getEngine();
+        webEngine.getLoadWorker().stateProperty().addListener((observable, oldValue, newValue) -> {
+            if (newValue == Worker.State.SUCCEEDED) {
```
Use static import here.

dheerajwebonise commented on src/main/java/com/webonise/views/MainScreenView.java:
```diff
@@ -0,0 +1,118 @@
+    @FXML
+    private TableColumn lngColumn;
+
+    public MainScreenView() {
```
Do we need to provide default constructor here?

dheerajwebonise commented on src/main/java/com/webonise/views/MainScreenView.java:
```diff
@@ -0,0 +1,118 @@
+    private void loadWebEngine() {
+        LOG.debug("Initializing webEngine");
+        webEngine = mapBrowser.getEngine();
+        webEngine.getLoadWorker().stateProperty().addListener((observable, oldValue, newValue) -> {
```
Use lambda here

dheerajwebonise commented on src/main/java/com/webonise/views/MainScreenView.java:
```diff
@@ -0,0 +1,118 @@
+import javax.annotation.PostConstruct;
+
+@Component
+public class MainScreenView extends AbstractScreenView {
```
favor composition over inheritance
http://stackoverflow.com/questions/49002/prefer-composition-over-inheritance

dheerajwebonise commented on src/main/java/com/webonise/views/MainScreenView.java:
```diff
@@ -0,0 +1,118 @@
+    private TableColumn lngColumn;
+
+    public MainScreenView() {
+        super();
```
Now i understand why we need this constructor

dheerajwebonise commented on src/main/resources/js/mapbox_script.js:
```diff
@@ -0,0 +1,74 @@
+
+function createPolygon(){
+    LOG.logDebug("Creating polygon connecting the coordinates");
+    if(polygon)
```
should provide scope for if statement

dheerajwebonise commented on src/main/resources/js/mapbox_script.js:
```diff
@@ -0,0 +1,74 @@
+
+function deleteMarkers(){
+    LOG.logInfo("Deleting all markers ");
+    map.eachLayer(function(layer) {
```
Nice!

murtaza-webonise commented on src/main/resources/js/mapbox_script.js:
```diff
@@ -0,0 +1,74 @@
+	map.on('click', function(event){
+		addToList(event.latlng);
+	});
+    markerCount=1;
```
please look into the formatting of javascript plz plz plz !!1
